### PR TITLE
Added ability to return more than one metric from a function

### LINKED ISSUE #3905: New Feature: Add ability to return more than one metric from metric function
Following discussion in gitter:

Add ability to return dict from metric function. Would be useful for e.g. confusion matrix. Proposed behavior

`r = f(y_true,y_pred)`

1. If `r` is a dict - report every `(key, value)` pair as metric with name `key`
2. Report  `r`  as metric with `f.__name__` name otherwise

### PR DESCRIPTION
**Pull request by kilotaras**
Resolves #3905

### CODE REVIEW

File: docs/templates/getting-started/sequential-model-guide.md [hunk 1]
```diff
@@ -137,6 +137,22 @@ model.compile(optimizer='rmsprop',
...
+def mean_pred(y_true, y_pred):
+    return K.mean(y_pred)
+
+def multi_metric(y_true, y_pred):
+    m1 = ...
+    m2 = ...
+    return {
+        'my_metric': m1,
+        'my_other_metric': m2
+    }
```
Comment (fchollet) [inline docs/templates/getting-started/sequential-model-guide.md]: In general I would prefer examples to be meaningful rather than dummy placeholders. It's a way to both explain usage patterns and convey a sense of what you can do. Also examples should include their import statements (`K` in this cas).

File: keras/engine/training.py [hunk 2]
```diff
@@ -682,6 +688,12 @@ def compile(self, optimizer, loss, metrics=[], loss_weights=None,
 
         self._collected_trainable_weights = collect_trainable_weights(self)
 
+    def _append_metric_name(self, layer_num, metric_name):
```
Comment (fchollet) [inline keras/engine/training.py]: Since this is ad-hoc and specific to a single method it shouldn't be a class method but rather a function defined on the fly in `compile`.

File: keras/engine/training.py [hunk 3]
```diff
@@ -644,27 +647,30 @@ def compile(self, optimizer, loss, metrics=[], loss_weights=None,
...
+                    self.metrics_tensors.append(acc_fn(y_true, y_pred))
+                    self._append_metric_name(i, 'acc')
                 else:
                     metric_fn = metrics_module.get(metric)
-                    self.metrics_tensors.append(metric_fn(y_true, y_pred))
-                    if len(self.output_names) == 1:
-                        self.metrics_names.append(metric_fn.__name__)
-                    else:
-                        self.metrics_names.append(self.output_layers[i].name + '_' + metric_fn.__name__)
+                    res = metric_fn(y_true, y_pred)
```
Comment (fchollet) [inline keras/engine/training.py]: Prefer more explicit variable names.

File: docs/templates/getting-started/sequential-model-guide.md [hunk 4]
```diff
@@ -137,6 +137,24 @@ model.compile(optimizer='rmsprop',
...
+def mean_pred(y_true, y_pred):
+    return K.mean(y_pred)
+
+def precision_recall(y_true, y_pred):
+    precision = ...
+    recall = ...
+    return {
+        'precision': precision,
+        'recall': recall,
+    }
```
Comment (fchollet) [inline docs/templates/getting-started/sequential-model-guide.md]: Precision and recall cannot be computed in a batched way, they are global metrics.

Comment (kilotaras) [inline docs/templates/getting-started/sequential-model-guide.md]: @fchollet what would you recommend? Maybe `binary_confusion_matrix`?

Comment (fchollet) [inline docs/templates/getting-started/sequential-model-guide.md]: Maybe mean incertitude of the predictions (distance to 0.5, if balanced binary classification problem). Maybe rate of false positives and false negatives. That last one can be pretty useful.

**Comment by fchollet** [comment]
Please resolve conflicts so we can merge. The PR is LGTM otherwise.

**Comment by kilotaras** [comment]
@fchollet I've fixed the conflicts

**Comment by fchollet** [comment]
Thanks!
